### PR TITLE
Capture dynamic qubit allocation v2

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

This PR Replaces #7673 

**Description of the Change:**

Adds allocate_prim and deallocate_prim that generate and consume dynamic wires (abstract ints) respectively.

**Benefits:**

Can capture qubit allocation for integration with catalyst.

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-93191]

### CODE REVIEW

File: pennylane/allocation.py [hunk 1]
```diff
@@ -159,7 +207,10 @@ def __enter__(self):
         return self
 
     def __exit__(self, *_, **__):
-        Deallocate(self)
+        if capture_enabled():
+            deallocate(self)
+        else:
+            Deallocate(self)
```
Comment (albi3ro) [inline pennylane/allocation.py]:
```suggestion
        deallocate(self)
```

File: pennylane/allocation.py [hunk 2]
```diff
@@ -15,11 +15,55 @@
...
 from pennylane.wires import Wires
 
+try:
+    import jax
+except ImportError:
+    jax = None
+
+
+@lru_cache()
+def _get_allocate_prim():
```
Comment (paul0403) [inline pennylane/allocation.py]: Is it possible to export the primitives themselves? In the plxpr to catalyst jaxpr conversion I have to discover the primitives through 
```python
@PLxPRToQuantumJaxprInterpreter.register_primitive(qml.allocation._get_allocate_prim())
```
which (a) feels awkward, and (b) relies on the (accidental?) implementation detail that each local `allocate_prim` variable is cached so they all happen to be the same one. By contrast, for the other primitives we just do something like 
```python
@PLxPRToQuantumJaxprInterpreter.register_primitive(qml.QubitUnitary._primitive)
```

File: pennylane/allocation.py [hunk 3]
```diff
@@ -15,11 +15,55 @@
...
+
+    # pylint: disable=unused-argument
+    @allocate_prim.def_impl
+    def _(*, num_wires, require_zeros=True, restored=False):
+        raise NotImplementedError("jaxpr containing qubit allocation cannot be executed.")
+
+    # pylint: disable=unused-argument
+    @allocate_prim.def_abstract_eval
+    def _(*, num_wires, require_zeros=True, restored=False):
+        return [jax.core.ShapedArray((), dtype=int) for _ in range(num_wires)]
```
Comment (paul0403) [inline pennylane/allocation.py:47]: Checking my understanding: the return values of the allocate primitive are the new wire indices? And then gates can just use them like any other wire index?

In other words, am I correct to understand that plxpr maintains a global list of wire indices, even across registers (I guess "register" is not a real thing in plxpr (yet?), but here I just mean a group of wires from an allocation).

**Review by paul0403 — APPROVED** [review]
Looks good! Just a small QoL request from me, and a question to check my plxpr understanding a bit :100:

File: tests/test_allocation.py [hunk 4]
```diff
@@ -193,3 +195,84 @@ def test_allocate_context_manager():
...
+            assert v.aval.dtype == jax.numpy.int64
+
+        assert jaxpr.eqns[1].invars[0] is jaxpr.eqns[0].outvars[0]
+
+        assert jaxpr.eqns[2].primitive == _get_deallocate_prim()
+        assert jaxpr.eqns[2].params == {}
+        assert jaxpr.eqns[2].invars == jaxpr.eqns[0].outvars
+
+    def test_no_implementation(self):
+        """Test that deallocation has no concrete implementation."""
```
Comment (dwierichs) [inline tests/test_allocation.py]:
```suggestion
        """Test that (de)allocation has no concrete implementation."""
```

**Review by dwierichs — APPROVED** [review]
LGTM! 
If that one suggestion is causing CI work, feel free to discard it!

File: pennylane/allocation.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/allocation.py:47]: Yep. They are abstract indices just like any other wire label.  We do not have a concept of a register yet in pennylane.  Wires are just integers with program capture.

File: pennylane/allocation.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/allocation.py]: Now you can access them as `qml.allocation.allocate_prim` and `qml.allocation.deallocate_prim`.  They'll just be `None` if jax isn't installed.
